File: src/main/java/org/hisp/dhis/expression/spi/AggregateMath.java
package org.hisp.dhis.expression.spi;

import java.util.Arrays;
import java.util.stream.DoubleStream;

/**
 * Math operations for aggregation functions.
 *
 * Some computations were extracted and simplified from apache math.
 *
 * @author Jan Bernitt
 */
final class AggregateMath {

    private AggregateMath() {
        throw new UnsupportedOperationException("util");
    }

    static double avg(double[] values) {
        double sampleSize = values.length;
        double xbar = sum(values) / sampleSize;
        double correction = 0.0d;

        for (double value : values) {
            correction += value - xbar;
        }

        return xbar + correction / sampleSize;
    }

    static double count(double[] values) {
        return values.length;
    }

    /**
     * Vector maximum.
     *
     * @param values input vector
     * @return largest of the values ignoring NaNs or NaN if values has length zero
     */
    static double max(double[] values) {
        return DoubleStream.of(values).filter(v -> !Double.isNaN(v)).max().orElse(Double.NaN);
    }

    static double median(double[] values) {
        return 0d;
    }

    /**
     * Vector minimum.
     *
     * @param values input vector
     * @return smallest of the values ignoring NaNs or NaN if values has length zero
     */

    static double min(double[] values) {
        return DoubleStream.of(values).filter(v -> !Double.isNaN(v)).min().orElse(Double.NaN);
    }

    /**
     * Vector function: percentileCont (continuous percentile)
     * <p/>
     * The percentileCont function is equivalent to the PostgreSQL function
     * percentile_cont and the Excel function PERCENTILE.INC
     *
     * @author Jim Grace
     */
    static Double percentileCont(double[] values, Number fraction) {
        if (values.length == 0 || fraction == null || fraction.doubleValue() < 0d || fraction.doubleValue() > 1d )
        {
            return null;
        }
        Arrays.sort( values );
        if ( fraction.doubleValue() == 0d )
        {
            return values[0];
        }
        double[] work = DoubleStream.of(values).filter(v -> !Double.isNaN(v)).toArray();
        if (work.length == 0) {
            return Double.NaN; // TODO make null to be consistent?
        }
        int[] pivotsHeap = new int[512];
        Arrays.fill(pivotsHeap, -1);
        double pos = indexR7(fraction.doubleValue(), work.length);
        return estimate(work, pivotsHeap, pos);
    }

    static double stddev(double[] values) {
        return 0d;
    }

    static double stddevPop(double[] values) {
        return 0d;
    }

    static double stddevSamp(double[] values) {
        return 0d;
    }

    /**
     * Vector sum.
     *
     * @param values input vector
     * @return sum of all values
     */
    static double sum(double[] values) {
        return DoubleStream.of(values).sum();
    }

    static double variance(double[] values) {
        return 0d;
    }

    private static double estimate(double[] work, int[] pivotsHeap, double pos) {
        int length = work.length;
        double fpos = Math.floor(pos);
        int intPos = (int)fpos;
        double dif = pos - fpos;
        if (pos < 1.0) {
            return KthSelector.select(work, pivotsHeap, 0);
        } else if (pos >= length) {
            return KthSelector.select(work, pivotsHeap, length - 1);
        } else {
            double lower = KthSelector.select(work, pivotsHeap, intPos - 1);
            double upper = KthSelector.select(work, pivotsHeap, intPos);
            return lower + dif * (upper - lower);
        }
    }

    private static double indexR7(double p, int length) {
        return Double.compare(p, 0.0) == 0 ? 0.0 : (Double.compare(p, 1.0) == 0 ? (double)length : 1.0 + (length - 1) * p);
    }

    /**
     * Extracted from {@code org.apache.commons.math3.util.KthSelector} and
     * {@code org.apache.commons.math3.util.MedianOf3PivotingStrategy}.
     */
    static class KthSelector {

        static double select(double[] work, int[] pivotsHeap, int k) {
            int begin = 0;
            int end = work.length;
            int node = 0;
            boolean usePivotsHeap = pivotsHeap != null;

            while (end - begin > 15) {
                int pivot;
                if (usePivotsHeap && node < pivotsHeap.length && pivotsHeap[node] >= 0) {
                    pivot = pivotsHeap[node];
                } else {
                    pivot = partition(work, begin, end, pivotIndex(work, begin, end));
                    if (usePivotsHeap && node < pivotsHeap.length) {
                        pivotsHeap[node] = pivot;
                    }
                }

                if (k == pivot) {
                    return work[k];
                }

                if (k < pivot) {
                    end = pivot;
                    node = Math.min(2 * node + 1, usePivotsHeap ? pivotsHeap.length : pivot);
                } else {
                    begin = pivot + 1;
                    node = Math.min(2 * node + 2, usePivotsHeap ? pivotsHeap.length : end);
                }
            }

            Arrays.sort(work, begin, end);
            return work[k];
        }

        private static int partition(double[] work, int begin, int end, int pivot) {
            double value = work[pivot];
            work[pivot] = work[begin];
            int i = begin + 1;
            int j = end - 1;

            while (i < j) {
                while (i < j && work[j] > value) {
                    --j;
                }

                while (i < j && work[i] < value) {
                    ++i;
                }

                if (i < j) {
                    double tmp = work[i];
                    work[i++] = work[j];
                    work[j--] = tmp;
                }
            }

            if (i >= end || work[i] > value) {
                --i;
            }

            work[begin] = work[i];
            work[i] = value;
            return i;
        }

        static int pivotIndex(double[] work, int begin, int end) {
            int inclusiveEnd = end - 1;
            int middle = begin + (inclusiveEnd - begin) / 2;
            double wBegin = work[begin];
            double wMiddle = work[middle];
            double wEnd = work[inclusiveEnd];
            if (wBegin < wMiddle) {
                if (wMiddle < wEnd) {
                    return middle;
                } else {
                    return wBegin < wEnd ? inclusiveEnd : begin;
                }
            } else if (wBegin < wEnd) {
                return begin;
            } else {
                return wMiddle < wEnd ? inclusiveEnd : middle;
            }
        }
    }
}

File: src/jvmMain/java/org/hisp/dhis/lib/expression/math/AggregateMath.kt
package org.hisp.dhis.lib.expression.math

/**
 * Math operations for aggregation functions.
 *
 *
 * Some computations were extracted and simplified from apache math.
 *
 * @author Jan Bernitt
 */
object AggregateMath {

    fun avg(values: DoubleArray): Double {
        val sampleSize = values.size.toDouble()
        val xbar = sum(values) / sampleSize
        var correction = 0.0
        for (value in values) {
            correction += value - xbar
        }
        return xbar + correction / sampleSize
    }

    fun count(values: DoubleArray): Double {
        return values.size.toDouble()
    }

    /**
     * Vector maximum.
     *
     * @param values input vector
     * @return largest of the values ignoring NaNs or NaN if values has length zero
     */
    fun max(values: DoubleArray): Double {
        return values.filter { v: Double -> !v.isNaN() }.maxOrNull() ?: Double.NaN
    }

    fun median(values: DoubleArray?): Double {
        //TODO implement
        return 0.0
    }

    /**
     * Vector minimum.
     *
     * @param values input vector
     * @return smallest of the values ignoring NaNs or NaN if values has length zero
     */
    fun min(values: DoubleArray): Double {
        return values.filter { v: Double -> !v.isNaN() }.minOrNull() ?: Double.NaN
    }

    /**
     * Vector function: percentileCont (continuous percentile)
     *
     * The percentileCont function is equivalent to the PostgreSQL function percentile_cont and the Excel function
     * PERCENTILE.INC
     *
     * @author Jim Grace
     */
    fun percentileCont(values: DoubleArray, fraction: Number?): Double? {
        if (values.isEmpty() || fraction == null || fraction.toDouble() < 0.0 || fraction.toDouble() > 1.0) {
            return null
        }
        values.sort()
        if (fraction.toDouble() == 0.0) {
            return values[0]
        }
        val work = values.filter { v: Double -> !v.isNaN() } .toDoubleArray()
        if (work.isEmpty()) {
            return Double.NaN // TODO make null to be consistent?
        }
        val pivotsHeap = IntArray(512)
        pivotsHeap.fill( -1)
        val pos = indexR7(fraction.toDouble(), work.size)
        return estimate(work, pivotsHeap, pos)
    }

    fun stddev(values: DoubleArray?): Double {
        //TODO implement
        return 0.0
    }

    fun stddevPop(values: DoubleArray?): Double {
        //TODO implement
        return 0.0
    }

    fun stddevSamp(values: DoubleArray?): Double {
        //TODO implement
        return 0.0
    }

    /**
     * Vector sum.
     *
     * @param values input vector
     * @return sum of all values
     */
    fun sum(values: DoubleArray): Double {
        return values.sum()
    }

    fun variance(values: DoubleArray?): Double {
        //TODO implement
        return 0.0
    }

    private fun estimate(work: DoubleArray, pivotsHeap: IntArray, pos: Double): Double {
        val length = work.size
        val fpos = Math.floor(pos)
        val intPos = fpos.toInt()
        val dif = pos - fpos
        return if (pos < 1.0) {
            KthSelector.select(work, pivotsHeap, 0)
        }
        else if (pos >= length) {
            KthSelector.select(work, pivotsHeap, length - 1)
        }
        else {
            val lower = KthSelector.select(work, pivotsHeap, intPos - 1)
            val upper = KthSelector.select(work, pivotsHeap, intPos)
            lower + dif * (upper - lower)
        }
    }

    private fun indexR7(p: Double, length: Int): Double {
        return when (p) {
            0.0 -> 0.0
            1.0 -> length.toDouble()
            else -> 1.0 + (length - 1) * p
        }
    }

    /**
     * Extracted from `org.apache.commons.math3.util.KthSelector` and
     * `org.apache.commons.math3.util.MedianOf3PivotingStrategy`.
     */
    internal object KthSelector {
        fun select(work: DoubleArray, pivotsHeap: IntArray?, k: Int): Double {
            var begin = 0
            var end = work.size
            var node = 0
            val usePivotsHeap = pivotsHeap != null
            while (end - begin > 15) {
                var pivot: Int
                if (usePivotsHeap && node < pivotsHeap!!.size && pivotsHeap[node] >= 0) {
                    pivot = pivotsHeap[node]
                }
                else {
                    pivot = partition(work, begin, end, pivotIndex(work, begin, end))
                    if (usePivotsHeap && node < pivotsHeap!!.size) {
                        pivotsHeap[node] = pivot
                    }
                }
                if (k == pivot) {
                    return work[k]
                }
                if (k < pivot) {
                    end = pivot
                    node = Math.min(2 * node + 1, if (usePivotsHeap) pivotsHeap!!.size else pivot)
                }
                else {
                    begin = pivot + 1
                    node = Math.min(2 * node + 2, if (usePivotsHeap) pivotsHeap!!.size else end)
                }
            }
            work.sort(begin, end)
            return work[k]
        }

        private fun partition(work: DoubleArray, begin: Int, end: Int, pivot: Int): Int {
            val value = work[pivot]
            work[pivot] = work[begin]
            var i = begin + 1
            var j = end - 1
            while (i < j) {
                while (i < j && work[j] > value) {
                    --j
                }
                while (i < j && work[i] < value) {
                    ++i
                }
                if (i < j) {
                    val tmp = work[i]
                    work[i++] = work[j]
                    work[j--] = tmp
                }
            }
            if (i >= end || work[i] > value) {
                --i
            }
            work[begin] = work[i]
            work[i] = value
            return i
        }

        private fun pivotIndex(work: DoubleArray, begin: Int, end: Int): Int {
            val inclusiveEnd = end - 1
            val middle = begin + (inclusiveEnd - begin) / 2
            val wBegin = work[begin]
            val wMiddle = work[middle]
            val wEnd = work[inclusiveEnd]
            return if (wBegin < wMiddle) {
                if (wMiddle < wEnd) {
                    middle
                }
                else {
                    if (wBegin < wEnd) inclusiveEnd else begin
                }
            }
            else if (wBegin < wEnd) {
                begin
            }
            else {
                if (wMiddle < wEnd) inclusiveEnd else middle
            }
        }
    }
}
